### PR TITLE
Add Service Member model, migration, and handlers

### PR DESCRIPTION
**Pull request by amitch23**
## Description

This PR adds the model, migration and handlers for a service member.

## Reviewer Notes
* This is the first model that we're implementing an "is complete" check as a method call on the struct. Do we want to abstract this for any type of struct?

## Code Review Verification Steps

* [x] All tests pass.
* [x] Code follows the guidelines for [Logging](https://github.com/transcom/mymove/blob/master/docs/backend.md#logging)
* [x] The requirements listed in
 [Querying the Database Safely](https://github.com/transcom/mymove/blob/master/docs/backend.md#querying-the-database-safely)
 have been satisfied.
* [x] There are no [aXe](https://www.deque.com/products/aXe/) warnings for UI.
* [ ] This works in IE.
* [ ] Any new third party client side dependencies (e.g., login.gov, google analytics, CDN libraries, etc.) have been communicated to @willowbl00.
* [x] Request review from a member of a different team.
* [ ] (TRIAL) Have the Pivotal acceptance criteria been met for this change?

## References

* [Pivotal story](https://www.pivotaltracker.com/story/show/156435129) for this change

## Screenshots
Post to /service_members/ for logged in user:
screen shot 2018-04-06 at 5 32 47 pm

### CODE REVIEW

File: pkg/models/service_member.go [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+			result = NewInvalidServiceMemberResult(FetchErrorForbidden)
+		} else {
+			result = NewValidServiceMemberResult(serviceMember)
+		}
+	}
+
+	return result, err
+}
+
+// ValidateServiceMemberOwnership validates that a user has a serviceMember that exists
```
[the remark this answers is not in the document]
Comment (amitch23) [reply pkg/models/service_member.go]: done

File: pkg/models/service_member.go [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+	if s.FirstName == nil {
+		return false
+	}
+	// Do we required a middle initial?
+	if s.LastName == nil {
+		return false
+	}
+	if s.Telephone == nil {
+		return false
+	}
```
[the remark this answers is not in the document]
Comment (amitch23) [reply pkg/models/service_member.go]: done

File: pkg/models/service_member.go [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+		serviceMember: serviceMember,
+	}
+}
+
+// GetServiceMemberForUser returns a serviceMember only if it is allowed for the given user to access that serviceMember.
+// If the user is not authorized to access that serviceMember, it behaves as if no such serviceMember exists.
+func GetServiceMemberForUser(db *pop.Connection, userID uuid.UUID, id uuid.UUID) (ServiceMemberResult, error) {
+	var result ServiceMemberResult
+	var serviceMember ServiceMember
+	err := db.Find(&serviceMember, id)
```
[the remark this answers is not in the document]
Comment (amitch23) [reply pkg/models/service_member.go]: done

File: pkg/handlers/service_members.go [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+		h.logger.Error("DB Query", zap.Error(err))
+		response = servicememberop.NewShowServiceMemberInternalServerError()
+	} else if !serviceMemberResult.IsValid() {
+		switch errCode := serviceMemberResult.ErrorCode(); errCode {
+		case models.FetchErrorNotFound:
+			response = servicememberop.NewShowServiceMemberNotFound()
+		case models.FetchErrorForbidden:
+			response = servicememberop.NewShowServiceMemberForbidden()
+		default:
+			h.logger.Fatal("An error type has occurred that is unaccounted for in this case statement.")
```
Comment (stangah) [inline pkg/handlers/service_members.go]: There shouldn't be any `logger.Fatal` calls anymore, return an internal server error if you need a default case action

File: pkg/handlers/service_members.go [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+	var response middleware.Responder
+	// Get user id from context
+	user, err := models.GetUserFromRequest(h.db, params.HTTPRequest)
+	if err != nil {
+		response = servicememberop.NewPatchServiceMemberUnauthorized()
+		return response
+	}
+	serviceMemberID, err := uuid.FromString(params.ServiceMemberID.String())
+	if err != nil {
+		h.logger.Fatal("Invalid ServiceMemberID, this should never happen.")
```
Comment (stangah) [inline pkg/handlers/service_members.go]: You should return a bad request response, not use `logger.Fatal`

File: pkg/handlers/service_members.go [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+		h.logger.Error("DB Error checking on serviceMember validity", zap.Error(err))
+		response = servicememberop.NewPatchServiceMemberInternalServerError()
+	} else if !serviceMemberResult.IsValid() {
+		switch errCode := serviceMemberResult.ErrorCode(); errCode {
+		case models.FetchErrorNotFound:
+			response = servicememberop.NewPatchServiceMemberNotFound()
+		case models.FetchErrorForbidden:
+			response = servicememberop.NewPatchServiceMemberForbidden()
+		default:
+			h.logger.Fatal("An error type has occurred that is unaccounted for in this case statement.")
```
Comment (stangah) [inline pkg/handlers/service_members.go]: 500 response instead

File: pkg/handlers/service_members.go [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+	} else { // The given serviceMember does belong to the current user.
+		serviceMember := serviceMemberResult.ServiceMember()
+		payload := params.PatchServiceMemberPayload
+
+		verrs, err := serviceMember.PatchServiceMemberWithPayload(h.db, payload)
+
+		if verrs.HasAny() {
+			response = servicememberop.NewPatchServiceMemberBadRequest()
+		} else if err != nil {
+			h.logger.Error("DB doingi things", zap.Error(err))
```
Comment (stangah) [inline pkg/handlers/service_members.go]: This can probably go :p

File: pkg/handlers/service_members.go (hunk 4 above)
Comment (amitch23) [reply pkg/handlers/service_members.go]: done

File: pkg/handlers/service_members.go (hunk 5 above)
Comment (amitch23) [reply pkg/handlers/service_members.go]: done

File: pkg/handlers/service_members.go (hunk 6 above)
Comment (amitch23) [reply pkg/handlers/service_members.go]: done

File: pkg/handlers/service_members.go (hunk 7 above)
Comment (amitch23) [reply pkg/handlers/service_members.go]: done

**Pull request merged**